# HEARTSVG
HEARTSVG: a high-efficiency and robust method for identifying spatial expression patterns in large-scale spatial transcriptomic data
We proposed HEARTSVG, which builds upon serial correlation tests for rapidly and precisely detecting SVGs without relying on the underlying data-generative model or pre-defining spatial patterns.Furthermore, we developed the software HEARTSVG to provide an SVG auto-clustering module for predictions of spatial domain and visualizations, allowing users to discover novel biological phenomena. 

To be able to use the R package HEARTSVG successfully, we suggest manually installing the following packages and their dependent packages first: spatstat, Seurat, plotly, plot3D, TTR, stringr, fpp2, seasonal, cluster, poolr, reshape2, data.table, plyr, dplyr, ggplot2, MASS, EnvStats, BreakPoints, gprofiler2

```{r,warning=F,message=F,results='hide',error=F}
library(spatstat);library(Seurat);library(plot3D);library(plotly);library(TTR);
library(stringr);library(fpp2);library(seasonal);library(cluster);library(poolr);
library(reshape2);library(data.table);library(plyr);library(dplyr);library(ggplot2);
library(MASS);library(EnvStats);library(BreakPoints);library(gprofiler2)

```
## Required R modules
```{r ,warning=F}
R (>= 3.5.0)
```

## Dependencies
```{r ,warning=F}
fpp2 (>=2.4)
seasonal (>=1.9.0)
cluster (>=2.1.3)
poolr (>=1.1.1)
Seurat (4.1.1)
spatstat (>=2.3.4)
ggplot2 (>=3.4.2)
EnvStats (>=2.7.0)
BreakPoints (>=1.2)
plotly (>=4.10.0)
plot3D (>=1.4)
gprofiler2 (>=0.2.1)


```

## Installation

```{r}
library(devtools)
devtools::install_github('https://github.com/cz0316/HEARTSVG',force = T)
library(HEARTSVG)
```

## Operating systems tested on (HEARTSVG 1.1.0) :
macOS Ventura 13.4.1 <br>
Windows 10 <br>


## 2023/12 Update
We optimized the code to achieve reductions in runtime and memory usage, and the optimized function is named **'heartsvg_fast'**. However, due to an unknown error, we are unable to compile the 'heartsvg_fast' function into the R package. As a result, we are providing it separately (file_name: [heartsvg_fast.R](https://github.com/cz0316/HEARTSVG/blob/main/heartsvg_fast.R)). We are currently investigating the cause of the error.




## 0 Load data

We use the ST data from the Wu et al. study as an example to introduce the usage of the R package HEARTSVG. The article and data download link are: https://doi.org/10.1158/2159-8290.CD-21-0316; http://www.cancerdiversity.asia/scCRLM/.


```{r ,warning=F}
stc2.dt=readRDS(file = '~/HEARTSVG/data/HEARTSVG_demo.RDS')
dim(stc2.dt)
## [1] 4174 15429
```
If your data is of type seuratObject, we provide two functions, **filter.gene** and **seurat.trans** , for converting the data into a format suitable for HEARTSVG.
We use the function **filter.gene** to filter the non-expressed genes and low expression proportion genes. Then we use the function **seurat.trans** to transform the data type, from SeuratObject to a data.frame.

```{r}
## stc2=filter.gene(stc2,thr=0.01) # thr=0.01 represents the minimum expression proportion of each gene across all cells.
## dim(stc2)
## [1] 8812 4174

## stc2.dt=seurat.trans(stc2)
## dim(stc2.dt)
## [1] 4174 8814
```


## 1 Find SVG

The input data of the function **heartsvg** is a data.frame with dimensions n*(p+2) contains gene coordinates and gene expression counts. 

The first 2 columns should represent coordinates and their column names must be 'row' and 'col', respectively. The remaining p columns correspond to the expression levels of p genes. There are n rows, representing n spots.
```{r}
demo=stc2.dt[,1:1002]
demo[1:6,1:6]
##                    row col SAMD11 NOC2L HES4 ISG15
## AAACAACGAATAGTTC-1   0  16      0     0    1     1
## AAACAAGTATCTCCCA-1  50 102      0     2    0     2
## AAACAATCTACTAGCA-1   3  43      0     1    0     0
## AAACACCAATAACTGC-1  59  19      1     0   11     6
## AAACAGAGCGACTCCT-1  14  94      0     0    1     2
## AAACAGCTTTCAGAAG-1  43   9      0     0    0     2
```


```{r}
## find SVG
result=heartsvg(demo,scale=T)
head(result)
##       gene pval p_adj rank
## 729  RPS27    0     0    1
## 604   PIGR    0     0    2
## 746 S100A6    0     0    3
## 185  CSRP1    0     0    4
## 725  RPL11    0     0    5
## 732   RPS8    0     0    6
```

Regarding the parameter **'scale'**, the default value is T. This parameter is used to mitigate the impact of extreme values caused by technical factors. 

If you are using technologies such as 10X Visium, which produce data with a moderate level of sparsity and some noise, we recommend setting **scale=T**.

Conversely, if you are using technologies such as HDST, which generate highly sparse data with very low expression levels, we recommend setting **scale=F**. 

Additionally, if you have already performed noise reduction, normalization, or other preprocessing steps on your ST data, we also recommend setting **scale=F**.


## 2 Visualization
Different visualizations of top SVGs.

### 2.1 2D plot
```{r}
svg.2Dplot(data = demo,gene=result$gene[1:3],method = 'raw')
```

![image](https://user-images.githubusercontent.com/57090974/227206644-dfdc02ab-94c6-480d-81a3-f92cb0dcd8a3.png)
![image](https://user-images.githubusercontent.com/57090974/227206706-d284478f-4c75-43d8-9855-c99b65c694be.png)
![image](https://user-images.githubusercontent.com/57090974/227206753-c0825a0e-8c85-4beb-8c73-fc0971ca9469.png)



### 2.2 Marginal expression plot of SVG

The marginal expression of genes are obtained by the semi-pooling process.

```{r}
svg.Mplot(data=demo,gene=result$gene[1:3],method = 'raw')
```

![image](https://user-images.githubusercontent.com/57090974/227206902-da9df34a-803b-4e73-a503-b3e10cc5d23f.png)
![image](https://user-images.githubusercontent.com/57090974/227206963-d036d030-827c-4791-8df3-79aa6a7a9e19.png)
![image](https://user-images.githubusercontent.com/57090974/227207038-98177ad3-46d6-48c8-8f82-9bfb5145f8e4.png)



### 2.3 3D plot

#### 2.3.1 3D scatter plot

```{r}
svg.3Dplot(data=demo,gene = c('RPS8'),type = 'scatter')

```

![newplot](https://user-images.githubusercontent.com/57090974/227207527-221c9d5c-8a7b-47da-8bd7-1f2686dd60b0.png)


#### 2.3.1 3D surface plot

```{r}
svg.3Dplot(data=demo,gene=c('RPS27'),type = 'surface')

```

![image](https://user-images.githubusercontent.com/57090974/227207699-86fa6205-3293-40fe-bbc0-3c74d8b3345e.png)


## 3 Prediction of spatial domains

We offers an auto-clustering module for SVGs that facilitates further analyses, including predicting spatial domains, conducting functional analyses, and visualization based on the set of SVG files.  
We categorize SVGs based on their similarity in spatial expression and adopt a data-driven approach to estimate the number of spatially coherent regions in both SVG expression and histology.

### 3.1 clustering for SVGs
```{r}
clust_demo=svg.clust(data=demo,svg=result$gene[1:500],method = 'h')
table(clust_demo$cluster)
## 
##  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17 
## 403  16  59   6   1   1   2   1   3   1   1   1   1   1   1   1   1 

```

```{r}
km_demo=svg.clust(data=demo,svg=result$gene[1:500],method = 'k',n.c=6)
table(km_demo$cluster)
## 
##   1   2   3   4   5   6 
## 114 139 123  27  68  29
```


### 3.2 visualization of spatial domains

```{r}
svg.2Dplot(data=demo,gene=clust_demogene[clustdemogene[clust_democluster==1],method = 'mean',title = 'domain-1')
```
![image](https://user-images.githubusercontent.com/57090974/227209204-cb369b7f-273b-4568-bc21-5477bb2ed4f0.png)


```{r}
svg.2Dplot(data=demo,gene=clust_demogene[clustdemogene[clust_democluster==2],method = 'mean',title = 'domain-2')
```

cl1.en=svg.enrichment(svg=clust_demo$gene[clust_demo$cluster==1])

head(cl1.en)


## 4 Enrichment

We perform enrichment analysis of SVG sets based on the package gprofiler2.

```{r}
cl1.en=svg.enrichment(svg=clust_demogene[clustdemogene[clust_democluster==1])

head(cl1.en)
##     query significant      p_value term_size query_size intersection_size
## 1 query_1        TRUE 5.007490e-08      5861        405               174
## 2 query_1        TRUE 2.599282e-07      4088        405               132
## 3 query_1        TRUE 5.064712e-07      6348        405               181
## 4 query_1        TRUE 5.795109e-07      5657        405               166
## 5 query_1        TRUE 1.015270e-06      6446        405               182
## 6 query_1        TRUE 2.018340e-06      2664        405                95
##   precision     recall    term_id source
## 1 0.4296296 0.02968777 GO:0048856  GO:BP
## 2 0.3259259 0.03228963 GO:0042221  GO:BP
## 3 0.4469136 0.02851292 GO:0048518  GO:BP
## 4 0.4098765 0.02934418 GO:0048522  GO:BP
## 5 0.4493827 0.02823456 GO:0032502  GO:BP
## 6 0.2345679 0.03566066 GO:0010033  GO:BP
##                                   term_name effective_domain_size source_order
## 1          anatomical structure development                 21128        13923
## 2                      response to chemical                 21128        10676
## 3 positive regulation of biological process                 21128        13622
## 4   positive regulation of cellular process                 21128        13626
## 5                     developmental process                 21128         8073
## 6             response to organic substance                 21128         3939
##                              parents
## 1                         GO:0032502
## 2                         GO:0050896
## 3             GO:0008150, GO:0050789
## 4 GO:0009987, GO:0048518, GO:0050794
## 5                         GO:0008150
## 6                         GO:0042221

```

![Rplot](https://user-images.githubusercontent.com/57090974/227210739-74e236d1-ab4b-4032-9b92-50565af266ad.png)


## Citation
Yuan, X., Ma, Y., Gao, R. et al. HEARTSVG: a fast and accurate method for identifying spatially variable genes in large-scale spatial transcriptomics. Nat Commun 15, 5700 (2024). https://doi.org/10.1038/s41467-024-49846-1


doi: https://doi.org/10.1038/s41467-024-49846-1

---
# Morphology-aware SVG Detection using Spatial Neighborhood Features

## Improving HEARTSVG: A Proxy Feature Approach for Spatial Transcriptomics
## í”„ë¡ì‹œ ì¶”ê°€ë¥¼ í†µí•œ HEARTSVGì˜ í•œê³„ì  ë³´ì™„ ì œì•ˆ

Project by: Taeyi Kim (ê¹€íƒœì´)
Base Method: HEARTSVG

ğŸ“Œ Project Overview

ì´ í”„ë¡œì íŠ¸ëŠ” ëŒ€ê·œëª¨ ê³µê°„ ì „ì‚¬ì²´(Spatial Transcriptomics) ë°ì´í„° ë¶„ì„ íˆ´ì¸ HEARTSVGë¥¼ ì‹¬ì¸µ ë¶„ì„í•˜ê³ , ê¸°ì¡´ ì•Œê³ ë¦¬ì¦˜ì˜ í•œê³„ë¥¼ ë³´ì™„í•˜ëŠ” ìƒˆë¡œìš´ ë°©ë²•ë¡ (Morphology-aware Proxy Features)ì„ ì œì•ˆ ë° ê²€ì¦í•œ ì—°êµ¬ì…ë‹ˆë‹¤.

ì—°êµ¬ëŠ” ë‹¤ìŒ ì„¸ ë‹¨ê³„ë¡œ ì§„í–‰ë˜ì—ˆìŠµë‹ˆë‹¤:

Paper Review: HEARTSVG ë…¼ë¬¸ ë¶„ì„ ë° ê¸°ì¡´ ë°©ë²•ë¡  ë¹„êµ

Code Analysis: ì•Œê³ ë¦¬ì¦˜ ë¡œì§ ë¶„ì„ ë° ë°ëª¨ ë°ì´í„° êµ¬ë™

Improvement Proposal (Core): í•œê³„ì  ë„ì¶œ ë° Proxy Feature(ì¡°ì§ë°€ì§‘ë„, ì—”íŠ¸ë¡œí”¼) ì œì•ˆ

1. Background & Limitations

1.1 HEARTSVG Overview

HEARTSVGëŠ” ë¹„ëª¨ìˆ˜ì (Distribution-free) í†µê³„ ê²€ì • ë°©ì‹ì„ ì‚¬ìš©í•˜ì—¬ ëŒ€ê·œëª¨ ê³µê°„ ì „ì‚¬ì²´ ë°ì´í„°ì—ì„œ **ê³µê°„ ë³€ì´ ìœ ì „ì(SVG)**ë¥¼ ë¹ ë¥´ê³  ì •í™•í•˜ê²Œ ì‹ë³„í•˜ëŠ” ì•Œê³ ë¦¬ì¦˜ì…ë‹ˆë‹¤.

1.2 Problem Definition (í•œê³„ì )

HEARTSVGëŠ” ê³µê°„ ì¢Œí‘œ($x, y$)ì™€ ë°œí˜„ëŸ‰($e$)ë§Œì„ ë³€ìˆ˜ë¡œ ì‚¬ìš©í•©ë‹ˆë‹¤. ì´ë¡œ ì¸í•´ ë‹¤ìŒê³¼ ê°™ì€ í•œê³„ê°€ ì¡´ì¬í•©ë‹ˆë‹¤.

í˜•íƒœí•™ì  ì •ë³´(Morphology) ë¶€ì¬: ì¡°ì§ì˜ ë°€ë„ë‚˜ ê²½ê³„ì„  ê°™ì€ ë³‘ë¦¬í•™ì  íŠ¹ì§•ì„ ë°˜ì˜í•˜ì§€ ëª»í•¨.

ë‹¨ìˆœ êµ¬ì¡°ì  ë§ˆì»¤ í¸í–¥: ë‹¨ìˆœíˆ ì„¸í¬ê°€ ë§ì´ ë­‰ì³ìˆëŠ” ê³³ì˜ ìœ ì „ì(ì˜ˆ: ë¯¸í† ì½˜ë“œë¦¬ì•„ ìœ ì „ì, Housekeeping gene)ê°€ ìµœìƒìœ„ ë­í¬(Rank 1~10)ë¥¼ ì°¨ì§€í•˜ëŠ” ê²½í–¥ì´ ìˆìŒ.

Disease-relevant Gene ëˆ„ë½: ì •ì‘ ì•”ì˜ ì¦ì‹ì´ë‚˜ ì „ì´ì™€ ê´€ë ¨ëœ í•µì‹¬ ìœ ì „ìëŠ” ë°œí˜„ëŸ‰ì´ ë‚®ê±°ë‚˜ êµ­ì†Œì ì´ì–´ì„œ í•˜ìœ„ê¶Œìœ¼ë¡œ ë°€ë ¤ë‚˜ëŠ” ë¬¸ì œ ë°œìƒ.

2. Proposed Method: Proxy Features

H&E ì—¼ìƒ‰ ì´ë¯¸ì§€ì™€ ê°™ì€ ì™¸ë¶€ ë°ì´í„° ì—†ì´, ì˜¤ì§ ì¢Œí‘œì™€ ë°œí˜„ ë°ì´í„°ë§Œìœ¼ë¡œ ì¡°ì§ì˜ í˜•íƒœí•™ì  íŠ¹ì§•ì„ ì¶”ì •í•˜ëŠ” ë‘ ê°€ì§€ **'Proxy Feature'**ë¥¼ ê³ ì•ˆí•˜ì—¬ ê¸°ì¡´ ì•Œê³ ë¦¬ì¦˜ì— ê°€ì¤‘ì¹˜ë¡œ ì ìš©í–ˆìŠµë‹ˆë‹¤.

ğŸ› ï¸ Method 1: Local Density Score (ì¡°ì§ ë°€ì§‘ë„)

ê°€ì„¤: ì•” ì¡°ì§ì€ ì„¸í¬ ë¶„ì—´ì´ í™œë°œí•˜ì—¬ ì •ìƒ ì¡°ì§ë³´ë‹¤ ì„¸í¬ ë°€ë„ê°€ ë†’ë‹¤. ë”°ë¼ì„œ ë°€ë„ê°€ ë†’ì€ ì˜ì—­ì˜ ìœ ì „ìëŠ” ì•”ì˜ ì¦ì‹/ìƒì¡´ê³¼ ì§ê²°ë  ê²ƒì´ë‹¤.

êµ¬í˜„: $k$-NN ($k=10$) ì•Œê³ ë¦¬ì¦˜ì„ ì´ìš©í•´ ì´ì›ƒ ê°„ í‰ê·  ê±°ë¦¬ì˜ ì—­ìˆ˜ë¡œ ë°€ë„ ì ìˆ˜ ì‚°ì¶œ.


$$Density \propto \frac{1}{\text{mean(distance to k-neighbors)}}$$

ğŸ› ï¸ Method 2: Neighborhood Entropy (ì´ì›ƒ ë‹¤ì–‘ì„±)

ê°€ì„¤: ì•”ì„¸í¬ì™€ ì •ìƒ ì„¸í¬ê°€ ë§ë‹¿ëŠ” **'ì¢…ì–‘ ê²½ê³„ì„ '**ì´ë‚˜ **'ë¯¸ì„¸í™˜ê²½(TME)'**ì€ ë‹¤ì–‘í•œ ì„¸í¬ê°€ í˜¼ì¬ë˜ì–´ ìˆë‹¤. ë³µì¡ì„±ì´ ë†’ì€ ê³³ì˜ ìœ ì „ìëŠ” ë©´ì—­ ë°˜ì‘/ì¹¨ìœ¤ê³¼ ê´€ë ¨ì´ ê¹Šë‹¤.

êµ¬í˜„: ë¹„ì§€ë„ í•™ìŠµ(K-means)ìœ¼ë¡œ ê°€ìƒì˜ ë¼ë²¨ì„ ë¶€ì—¬í•œ ë’¤, ì£¼ë³€ ì´ì›ƒ($k=20$)ì˜ Shannon Entropy ê³„ì‚°.

3. Experiment Results

ë…¼ë¬¸ì—ì„œ ì‚¬ìš©ëœ ëŒ€ì¥ì•”(Colorectal Cancer, CRC) ë°ì´í„°ë¥¼ ì‚¬ìš©í•˜ì—¬ Baseline(ê¸°ì¡´)ê³¼ ì œì•ˆ ëª¨ë¸(Proxy ì ìš©)ì„ ë¹„êµ ë¶„ì„í–ˆìŠµë‹ˆë‹¤.

ğŸ“Š 1. Baseline (HEARTSVG Original)

Top Rank: MT-ATP6, MT-CO2 (ë¯¸í† ì½˜ë“œë¦¬ì•„ ìœ ì „ì), B2M (6ìœ„)

í•´ì„: ë‹¨ìˆœíˆ ì„¸í¬ê°€ ì¡´ì¬í•˜ëŠ” ìœ„ì¹˜ë‚˜ êµ¬ì¡°ì  íŠ¹ì§•ë§Œì„ í¬ì°©í•¨. ë³‘ë¦¬í•™ì  ì˜ë¯¸ê°€ ë¶€ì¡±í•¨.

ğŸ“Š 2. Improvement with Density Proxy

Top Rank: LGR5 (4ìœ„), UHRF1 (2ìœ„)

ì„±ê³¼:

ê¸°ì¡´ 3,042ìœ„ì˜€ë˜ ëŒ€ì¥ì•” ì¤„ê¸°ì„¸í¬ ë§ˆì»¤ **LGR5**ë¥¼ 4ìœ„ë¡œ ê¸‰ìƒìŠ¹ì‹œí‚´.

ì•”ì„¸í¬ ì¦ì‹ ë§ˆì»¤ UHRF1ì„ 2ìœ„ë¡œ ì‹ë³„.

ì˜ì˜: ì•”ì„¸í¬ê°€ ê°€ì¥ ë¹½ë¹½í•˜ê²Œ ë­‰ì³ìˆëŠ” **'ì¢…ì–‘ í•µì‹¬ë¶€(Tumor Core)'**ë¥¼ ì •í™•íˆ íƒ€ê²©í•¨.

ğŸ“Š 3. Improvement with Entropy Proxy

Top Rank: IGLV6-57 (6ìœ„), COL10A1, MMP7

ì„±ê³¼:

ë©´ì—­ ìœ ì „ì(IGLV)ì™€ ê¸°ì§ˆ/ì¹¨ìœ¤ íš¨ì†Œ(COL, MMP)ê°€ ìƒìœ„ê¶Œ ì§„ì….

ì¡°ì§ ì „ë°˜ì— ê· ì¼í•œ B2Mì€ 6ìœ„ $\to$ 5,982ìœ„ë¡œ í•˜ë½ (ë‹¨ìˆœ ë°°ê²½ ì‹ í˜¸ í•„í„°ë§).

ì˜ì˜: ì•”ì„¸í¬ì™€ ì£¼ë³€ í™˜ê²½ì´ ìƒí˜¸ì‘ìš©í•˜ëŠ” **'ì¢…ì–‘ ë¯¸ì„¸í™˜ê²½(TME) ë° ê²½ê³„ì„ '**ì„ ì„±ê³µì ìœ¼ë¡œ í¬ì°©í•¨.

ğŸ–¼ï¸ Visualization Comparison

Baseline (Noise/Structure)

Density Proxy (Tumor Core)

Entropy Proxy (Boundary/TME)







MT-ATP6 (Rank 1)

LGR5 (Rank 4)

DSCR8 / IGLV6-57

4. Conclusion

ë³¸ ì—°êµ¬ëŠ” ê³ ê°€ì˜ ì´ë¯¸ì§€ ë°ì´í„° ì²˜ë¦¬ ê³¼ì • ì—†ì´, **ì¢Œí‘œ ê¸°ë°˜ì˜ ìˆ˜ì¹˜ì  í”„ë¡ì‹œ(Density, Entropy)**ë§Œìœ¼ë¡œë„ HEARTSVGì˜ ìƒë¬¼í•™ì  ë¯¼ê°ë„ë¥¼ íšê¸°ì ìœ¼ë¡œ ê°œì„ í•  ìˆ˜ ìˆìŒì„ ì…ì¦í–ˆìŠµë‹ˆë‹¤.

Density Proxy: ì•”ì˜ ì„±ì¥ê³¼ ì¦ì‹(Core) ê·œëª… íŠ¹í™”

Entropy Proxy: ì•”ì˜ ì „ì´ì™€ ë©´ì—­ ë°˜ì‘(Boundary) ê·œëª… íŠ¹í™”

ì´ëŠ” í–¥í›„ ëŒ€ê·œëª¨ ê³µê°„ ì „ì‚¬ì²´ ë°ì´í„°ë¥¼ ì €ë¹„ìš©ìœ¼ë¡œ ì‹ ì†í•˜ê²Œ ë¶„ì„í•˜ì—¬ ì§ˆë³‘ì˜ ë‹¨ê³„ì™€ íŠ¹ì„±ì„ ì…ì²´ì ìœ¼ë¡œ íŒŒì•…í•˜ëŠ” ìƒˆë¡œìš´ ë¶„ì„ í”„ë ˆì„ì›Œí¬ê°€ ë  ìˆ˜ ìˆìŠµë‹ˆë‹¤.

